File: c/indexof.c
/* Streaming search for a needle in a file, 64KiB chunks.
 * Handles needles that straddle chunk boundaries by carrying over
 * the last (needle_len - 1) bytes of each chunk.
 *
 * Usage: ./indexof [--mem] [file]
 *   --mem: load the first 256MiB into memory and time the search alone
 *          (isolates the memmem cost from I/O and process startup).
 */
#define _GNU_SOURCE /* memmem */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define CHUNK 65536
#define MEM_BYTES (256L * 1024 * 1024)
#define MEM_RUNS 10

static const char NEEDLE[] = "--boundary--";
#define NEEDLE_LEN (sizeof(NEEDLE) - 1)

static double now_sec(void)
{
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec + ts.tv_nsec / 1e9;
}

static int mem_bench(FILE *file)
{
    char *data = malloc(MEM_BYTES);
    if (data == NULL)
    {
        fprintf(stderr, "out of memory\n");
        return 1;
    }
    size_t size = fread(data, 1, MEM_BYTES, file);

    const char *p = NULL;
    double start = now_sec();
    for (int i = 0; i < MEM_RUNS; i++)
        p = memmem(data, size, NEEDLE, NEEDLE_LEN);
    double avg = (now_sec() - start) / MEM_RUNS;

    long index = p ? (long)(p - data) : -1;
    double gibs = size / avg / (1024.0 * 1024.0 * 1024.0);
    printf("mem-search bytes=%zu runs=%d avg_ms=%.2f throughput_gib_s=%.2f index=%ld\n",
           size, MEM_RUNS, avg * 1000, gibs, index);
    free(data);
    return 0;
}

static int stream_bench(FILE *file)
{
    static char buf[NEEDLE_LEN - 1 + CHUNK];
    size_t tail = 0;         /* carried-over bytes at the start of buf */
    long long offset = 0;    /* global index of buf[0] */
    long long chunks = 0;
    size_t n;
    while ((n = fread(buf + tail, 1, CHUNK, file)) > 0)
    {
        chunks++;
        size_t total = tail + n;
        char *p = memmem(buf, total, NEEDLE, NEEDLE_LEN);
        if (p != NULL)
        {
            printf("Number of chunks: %lld\n", chunks);
            printf("Found index at: %lld\n", offset + (p - buf));
            return 0;
        }
        size_t keep = total < NEEDLE_LEN - 1 ? total : NEEDLE_LEN - 1;
        memmove(buf, buf + total - keep, keep);
        offset += total - keep;
        tail = keep;
    }
    return 0;
}

int main(int argc, char **argv)
{
    int mem_mode = 0;
    const char *path = "../sample_file/file.bin";
    for (int i = 1; i < argc; i++)
    {
        if (strcmp(argv[i], "--mem") == 0)
            mem_mode = 1;
        else
            path = argv[i];
    }

    FILE *file = fopen(path, "rb");
    if (file == NULL)
    {
        fprintf(stderr, "cannot read file %s\n", path);
        return 1;
    }
    int rc = mem_mode ? mem_bench(file) : stream_bench(file);
    fclose(file);
    return rc;
}

File: go/indexof.go
// Streaming search for a needle in a file, 64KiB chunks.
// Handles needles that straddle chunk boundaries by carrying over
// the last (len(needle) - 1) bytes of each chunk.
//
// Usage: ./indexof [--mem] [file]
//
//	--mem: load the first 256MiB into memory and time the search alone
//	       (isolates the bytes.Index cost from I/O and process startup).
package main

import (
	"bytes"
	"fmt"
	"io"
	"os"
	"time"
)

const (
	chunkSize = 65536
	memBytes  = 256 * 1024 * 1024
	memRuns   = 10
)

var needle = []byte("--boundary--")

func memBench(path string) error {
	f, err := os.Open(path)
	if err != nil {
		return err
	}
	defer f.Close()

	info, err := f.Stat()
	if err != nil {
		return err
	}
	size := info.Size()
	if size > memBytes {
		size = memBytes
	}
	data := make([]byte, size)
	if _, err := io.ReadFull(f, data); err != nil {
		return err
	}

	found := -1
	start := time.Now()
	for i := 0; i < memRuns; i++ {
		found = bytes.Index(data, needle)
	}
	elapsed := time.Since(start).Seconds()

	avg := elapsed / memRuns
	gibs := float64(size) / avg / (1 << 30)
	fmt.Printf("mem-search bytes=%d runs=%d avg_ms=%.2f throughput_gib_s=%.2f index=%d\n",
		size, memRuns, avg*1000, gibs, found)
	return nil
}

func streamBench(path string) error {
	f, err := os.Open(path)
	if err != nil {
		return err
	}
	defer f.Close()

	overlap := len(needle) - 1
	buf := make([]byte, overlap+chunkSize)
	tail := 0     // carried-over bytes at the start of buf
	offset := 0   // global index of buf[0]
	chunks := 0
	for {
		n, err := f.Read(buf[tail : tail+chunkSize])
		if n > 0 {
			chunks++
			hay := buf[:tail+n]
			if i := bytes.Index(hay, needle); i != -1 {
				fmt.Println("Number of chunks:", chunks)
				fmt.Println("Found index at:", offset+i)
				return nil
			}
			keep := overlap
			if len(hay) < keep {
				keep = len(hay)
			}
			copy(buf, hay[len(hay)-keep:])
			offset += len(hay) - keep
			tail = keep
		}
		if err == io.EOF {
			return nil
		}
		if err != nil {
			return err
		}
	}
}

func main() {
	memMode := false
	path := ""
	for _, a := range os.Args[1:] {
		if a == "--mem" {
			memMode = true
		} else {
			path = a
		}
	}
	if path == "" {
		path = "../sample_file/file.bin"
	}

	var err error
	if memMode {
		err = memBench(path)
	} else {
		err = streamBench(path)
	}
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
}
